samples/get_vm_power_state.py

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python
+# William Lam
+# www.virtuallyghetto.com
+
+"""
+vSphere Python SDK program for listing all ESXi datastores and their
+associated devices
+"""
+import argparse
+import atexit
+
+from pyVim import connect
+from pyVmomi import vmodl
+from pyVmomi import vim
+
+from tools import cli
+
+
+def get_args():
+    """
+    Supports the command-line arguments listed below.
+    """
+
+    parser = argparse.ArgumentParser(
+        description='Process args for retrieving all the Virtual Machines')
+    parser.add_argument('-s', '--host', required=True, action='store',
+                        help='Remote host to connect to')
+    parser.add_argument('-o', '--port', type=int, default=443, action='store',
+                        help='Port to connect on')
+    parser.add_argument('-u', '--user', required=True, action='store',
+                        help='User name to use when connecting to host')
+    parser.add_argument('-p', '--password', required=True, action='store',
+                        help='Password to use when connecting to host')
+    parser.add_argument('-v', '--vm-name', required=True, action='store',
+                        help='Name of VM to get state on')
+    parser.add_argument('-j', '--json', default=False, action='store_true',
+                        help='Output to JSON')
+    args = parser.parse_args()
+    return args
+
+
+def main():
+    """
+    Simple command-line program for listing all ESXi datastores and their
+    associated devices
+    """
+
+    args = get_args()
+
+    cli.prompt_for_password(args)
+
+    try:
+        service_instance = connect.SmartConnect(host=args.host,
+                                                user=args.user,
+                                                pwd=args.password,
+                                                port=int(args.port))
+        if not service_instance:
+            print("Could not connect to the specified host using specified "
+                  "username and password")
+            return -1
+
+        atexit.register(connect.Disconnect, service_instance)
+
+        content = service_instance.RetrieveContent()
+        # Search for all ESXi hosts
+        objview = content.viewManager.CreateContainerView(content.rootFolder,
+                                                          [vim.VirtualMachine],
+                                                          True)
+        vms = objview.view
+        objview.Destroy()
+
+        for vm in vms:
+            if vm.name == args.vm_name:
```

Comment:
What if you have thousands of VMs in the inventory? Why pull this for all of them for that one VM? We should be showing how to use the API in the most efficient ways possible. In this case something from the searchIndex would be way better so that you only return that one item you are looking for. Take a look at the findByxxx functions of SearchIndex for an example. Here are a couple of samples that use some of those methods: [here](https://github.com/vmware/pyvmomi-community-samples/blob/master/samples/add_vm_extra_config_tags.py#L56), or better: [here](https://github.com/vmware/pyvmomi-community-samples/blob/master/samples/reboot_vm.py#L59)